src/app/models/function.rb

```diff
@@ -6,11 +6,30 @@ class Function < ApplicationRecord
 
   belongs_to :user
 
+  before_validation :function_name
+
   def to_param
     name
   end
 
   def self.from_param(param)
     find_by(name: param)
   end
+
+  private
+
+  def walk_ast(ast)
+    return ast.children.first.to_s if ast.respond_to?('type') && ast.type == :DEFN
+
+    return unless ast.respond_to?('children')
+
+    ast.children.each do |child|
+      func_name = walk_ast(child)
+      return func_name unless func_name.nil?
+    end
+  end
+
+  def function_name
+    self.name = walk_ast(RubyVM::AbstractSyntaxTree.parse(code))
```

Comment:
[R] When the code is invalid it will raise an `SyntaxError` 
in this case we need to 
1. Add error to the code field
2. set the name to nil so if the user updated his code to invalid ruby code it doesn't retain the old name extracted from valid code

Also this error will appear when you have a test case for invalid code case